src/main/java/io/jenkins/plugins/checks/github/GitHubChecksContext.java

```diff
@@ -150,14 +157,32 @@ private String resolveHeadSha() {
                             getRepository(), head.get().getName()));
         }
 
-        if (revision.get() instanceof AbstractGitSCMSource.SCMRevisionImpl) {
-            return ((AbstractGitSCMSource.SCMRevisionImpl) revision.get()).getHash();
+        return resolveCommit(revision.get());
+    }
+
+    private String resolveCommit() {
+        if (run == null) {
```

Comment:
In my refactored version `run` is not a field anymore and thus cannot be `null`. Rather than setting `run` to `null` we should immediately initialize all fields in the constructor. Then we do not have any `null` checks for `run` anymore. 

I'm not sure if I did understand the GitHub REST API correctly: is it valid to call it without a SHA? If not then we should not create a ChecksPublisher. This is the way I did it with Freestyle jobs and pipelines.

Reply:
The reason we introduce the nullable run here is that when the job is still in the queue, the run is not available, so we can't get the url of the run (see https://github.com/jenkinsci/github-checks-plugin/pull/13#discussion_r464201246) and will return the url of the job view instead.

Reply:
I didn't find a better way to avoid nullable run, but maybe we can add another boolean field `isLastBuild` when constructing the context to indicate whether the `run` is the last build? So when `isLastBuild == true`, we get run by `job.getLastBuild()`, when `isLastBuild == false`, we use the run that was passed in the constructor.

Reply:
I saw your refactoring on this problem, simple but great!  Ignore the previous two comments :)

Reply:
> I'm not sure if I did understand the GitHub REST API correctly: is it valid to call it without a SHA? If not then we should not create a ChecksPublisher. This is the way I did it with Freestyle jobs and pipelines.

The SHA is required for a check, without it, the check cannot be created successfully.